### PR TITLE
feat: add entity metadata links

### LINKED ISSUE #3787: [RFC] Entity Links
**Status:** Open for comments

This is a formal proposal for a [discussion on discord](https://discord.com/channels/687207715902193673/687207715902193679/789554424300830720)

## Need

It is common to have existing external information or references related to entities in the catalog. These could often be represented as hyperlinks to some external reference when no related well known annotation or plugin exists. Some examples of this might be.

- Deployed website, demo site, etc.
- Admin dashboard for a service/website
- External documentation, API references, storybooks, etc.
- External team README page
- Runtime dashboards, etc.
- Various SasS integrations
- ... and the list goes on and on

Having the ability to attach this information to entities would open the world for providing a common `Links` card component on each entity page. These links could contain some basic context in order to indicate their semantic purpose (e.g. `admin-dashboard`, or `api-docs`, etc.)

image

## Proposal

Provide a lightweight and quick way to extend an entity with an optional list of basic links to external information & references. This new field could be support some well-known links, org specific links, or specific entity instance links. The list of links could then be managed by entity owners or possibly automatically appended through catalog processors.

```yaml
apiVersion: backstage.io/v1alpha1
kind: Component
metadata:
  name: petstore
  links: # optional
    - url: https://admin.petstore.com  # required
      rel: admin-dashboard # optional but recommended to provide semantic meaning
      label: Admin Dashboard # optional and could be inferred by rel if omitted
spec:
  type: website
  ...
```

It would also be nice to have the ability to associate icons to links. This could be done through the `rel` attribute or perhaps more appropriately an optional `icon` field (or both with `rel` as a fallback if provided). The icon value could be an enum value that maps to a icon library such as [iconify](https://iconify.design/).

```yaml
metadata:
  links:
    - url: https://admin.petstore.com
      label: Admin Dashboard
      icon: dashboard-icon
```

It might be possible to have a short form string version that might be expanded to it's constituent parts. However escaping might be a concern that makes this less feasible. For example:

```yaml
metdata:
  links:
    - http://admin.petstore.com # just a url
    - url=http://admin.petstore.com,rel=admin-dashboard # a inline map
    - { url: 'http://admin.petstore.com' } # inline yaml object
```

## Alternatives

Links could probably be modeled as annotations:

```yaml
metdata:
  annotations:
    # Format: [links.backstage.io/${rel}]: [value]
    links.backstage.io/admin-dashboard: https://admin.petstore.com
```

However, since annotations are typically single values the ability to provide additional context of a `label` & `icon`, etc. could be harder to maintain.

## Risks

The concept of links can overlap with the existing entity `annotations` and possibly `tags`. They also could be modeled by extended either of these existing concepts. Also the concept of `links` is quite subjective and could mean anything making it a bit abstract and difficult to define when you would use a `link` versus an `annotation`.

Forcing people to use `annotations` to model a simple external reference could mean they don't attach this useful information as it requires additional setup for a specific case. On the flip side having a quick to use `links` field could be used as a dumping ground and therefore deter people from modelling proper annotations when appropriate. However, these potential points of confusion can be reduced through guidance and documentation.

### PR DESCRIPTION
**Pull request by andrewthauer**
## Hey, I just made a Pull Request!

Partially addresses #3787

This introduces the base concepts for a new optional entity `metadata.links` field. Follow up PR's will introduce a common react catalog component to render entity links within entity pages.

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

### CODE REVIEW

File: packages/catalog-model/src/entity/policies/FieldFormatEntityPolicy.ts [hunk 1]
```diff
@@ -134,6 +134,12 @@ export class FieldFormatEntityPolicy implements EntityPolicy {
       require(`tags.${i}`, tags[i], this.validators.isValidTag);
     }
 
+    const links = entity.metadata.links ?? [];
+
+    for (let i = 0; i < links.length; ++i) {
+      optional(`links.${i}`, links[i]?.icon, this.validators.isValidEntityName);
```
Comment (andrewthauer) [inline packages/catalog-model/src/entity/policies/FieldFormatEntityPolicy.ts:140]: Wasn't sure if I should introduce a new check or re-use the entity name validation for the `icon` key. I also wasn't sure if I should add new `required` validation check for the `url`.

File: packages/catalog-model/examples/acme/org.yaml [hunk 2]
```diff
@@ -3,6 +3,9 @@ kind: Group
 metadata:
   name: acme-corp
   description: The acme-corp organization
+  links:
```
Comment (Fox32) [inline packages/catalog-model/examples/acme/org.yaml:6]: Maybe we should have at least one example with more than one link? Except I missed it 😁

File: packages/catalog-model/examples/apis/swapi-graphql.yaml [hunk 3]
```diff
@@ -3,1174 +3,13 @@ kind: API
...
-    """An edge in a connection."""
-    type VehiclesEdge {
-      """The item at the end of the edge"""
-      node: Vehicle
-
-      """A cursor for use in pagination"""
-      cursor: String!
-    }
+  definition:
+    $text: https://github.com/graphql/swapi-graphql/blob/master/schema.graphql
```
Comment (freben) [inline packages/catalog-model/examples/apis/swapi-graphql.yaml:15]: Hm, this is an unrelated change - drawback being, now you're subject to github rate limits instead of all data being local.

I'm actually OK with it though. Should work fine as long as we have the implicit github.com integration always being added whether it's in one's config or not.

File: packages/catalog-model/src/entity/policies/FieldFormatEntityPolicy.test.ts [hunk 4]
```diff
@@ -110,4 +114,9 @@ describe('FieldFormatEntityPolicy', () => {
     data.metadata.tags.push('Hello World');
     await expect(policy.enforce(data)).rejects.toThrow(/tags.*"Hello World"/i);
   });
+
```
Comment (freben) [inline packages/catalog-model/src/entity/policies/FieldFormatEntityPolicy.test.ts:117]: I'd like these tests to be very detailed and to exercise every single possible variation :)

- accepts missing links
- accepts empty links array
- rejects bad links value

- rejects missing links.url
- rejects empty links.url value
- rejects bad links.url value

- accepts missing links.label
- rejects empty links.label value
- rejects bad links.label value

- accepts missing links.icon
- rejects empty links.icon value
- rejects bad links.icon value

File: packages/catalog-model/src/entity/policies/FieldFormatEntityPolicy.ts (hunk 1 above)
Comment (freben) [reply packages/catalog-model/src/entity/policies/FieldFormatEntityPolicy.ts:140]: Yeah there should for sure be a required check for that, but that's supposed to be covered by the `SchemaValidEntityPolicy`. So this new field should be added here with the proper semantics -> https://github.com/backstage/backstage/blob/master/packages/catalog-model/src/entity/policies/SchemaValidEntityPolicy.ts#L21

File: packages/catalog-model/src/entity/Entity.ts [hunk 5]
```diff
@@ -161,3 +166,23 @@ export type EntityRelationSpec = {
    */
   target: EntityName;
 };
+
+/**
+ * A link to external information that is related to the entity.
+ */
+export type EntityLink = {
```
Comment (freben) [inline packages/catalog-model/src/entity/Entity.ts:173]: this should be exported in `index.ts` too

File: docs/features/software-catalog/descriptor-format.md [hunk 6]
```diff
@@ -314,6 +314,34 @@ This field is optional, and currently has no special semantics.
 Each tag must be sequences of `[a-z0-9]` separated by `-`, at most 63 characters
 in total.
 
+### `links` [optional]
```
Comment (freben) [inline docs/features/software-catalog/descriptor-format.md:317]: could you add this field to the example at the very top of the file? both in the json and yaml ones that are next to each other

File: packages/catalog-model/examples/apis/swapi-graphql.yaml (hunk 3 above)
Comment (andrewthauer) [reply packages/catalog-model/examples/apis/swapi-graphql.yaml:15]: Good point. I'll revert this change.

Comment (Fox32) [reply packages/catalog-model/examples/apis/swapi-graphql.yaml:15]: >Hm, this is an unrelated change - drawback being, now you're subject to github rate limits instead of all data being local.

In the end we are already doing it for most other API files, but I would prefer to keep this as-is to have another test case without `$text`.

File: docs/features/software-catalog/descriptor-format.md [hunk 7]
```diff
@@ -54,6 +54,11 @@ software catalog API.
     "labels": {
       "system": "public-websites"
     },
+    "links": [{
+      "url": "https://admin.example-org.com",
+      "title": "Admin Dashboard",
+      "icon": "dashboard"
+    }],
```
Comment (freben) [inline docs/features/software-catalog/descriptor-format.md:61]: in the yaml below too!

File: packages/catalog-model/examples/apis/swapi-graphql.yaml (hunk 3 above)
Comment (freben) [reply packages/catalog-model/examples/apis/swapi-graphql.yaml]: Ah of course, you are right. But OK, revert sounds fine anyway

File: packages/catalog-model/src/entity/policies/FieldFormatEntityPolicy.ts (hunk 1 above)
Comment (andrewthauer) [reply packages/catalog-model/src/entity/policies/FieldFormatEntityPolicy.ts]: Added additional validations and tests :)

File: packages/catalog-model/src/entity/policies/FieldFormatEntityPolicy.test.ts (hunk 4 above)
Comment (andrewthauer) [reply packages/catalog-model/src/entity/policies/FieldFormatEntityPolicy.test.ts:117]: Done

File: packages/catalog-model/src/entity/Entity.ts (hunk 5 above)
Comment (andrewthauer) [reply packages/catalog-model/src/entity/Entity.ts:173]: Added

File: docs/features/software-catalog/descriptor-format.md (hunk 7 above)
Comment (andrewthauer) [reply docs/features/software-catalog/descriptor-format.md:61]: Whoops missed that. Added now.

**Comment by andrewthauer** [comment]
@freben, @Fox32 - I believe I've addressed all the comments. I feel slightly bad about the `isValidUrl` function using a try catch on `new URL(url)`. This could be switched over to a well crafted regex, shared library, or some other suggestion.

File: .changeset/fuzzy-pumpkins-tie.md [hunk 8]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/catalog-model': minor
```
Comment (andrewthauer) [inline .changeset/fuzzy-pumpkins-tie.md:2]: Maybe this should be patch since it's additive?

Comment (Fox32) [reply .changeset/fuzzy-pumpkins-tie.md:2]: It's also not a breaking change, so patch should be fine?

File: packages/catalog-model/src/validation/types.ts [hunk 9]
```diff
@@ -24,4 +24,7 @@ export type Validators = {
   isValidAnnotationKey(value: unknown): boolean;
   isValidAnnotationValue(value: unknown): boolean;
   isValidTag(value: unknown): boolean;
+  isValidLinkUrl(value: unknown): boolean;
+  isValidLinkTitle(value: unknown): boolean;
+  isValidLinkIcon(value: unknown): boolean;
```
Comment (freben) [inline packages/catalog-model/src/validation/types.ts:29]: This is an actual "breaking" change.

I'm not sure these are worth it. A valid url is a valid url - so having it in common is fine but maybe we can skip it here. And the title, the non-empty part of it is something we just typically slap in the schema because it makes absolutely no sense to allow the empty string. For the icons - well, not sure about that one, that's the only one where some marginal value could be had but, since it's just some nomenclature that people are free to choose as they see fit, and only within this very small scope of Backstage icons, I think the fixed rule of "same as metadata.name" is plenty forgiving.

Thoughts? I'd prefer just dropping these personally.

**Review by freben — APPROVED** [review]
A few little things remaining. Thank you for this work, almost there! <3

File: docs/features/software-catalog/descriptor-format.md [hunk 10]
```diff
@@ -81,6 +86,10 @@ metadata:
     circleci.com/project-slug: github/example-org/artist-website
   tags:
     - java
+  links:
+    url: https://admin.example-org.com
```
[the remark this answers is not in the document]
Comment (andrewthauer) [reply docs/features/software-catalog/descriptor-format.md:90]: Thanks for the catch 🤕

File: packages/catalog-model/examples/acme/org.yaml [hunk 11]
```diff
@@ -3,6 +3,11 @@ kind: Group
...
+    - url: http://www.acme.com/
+      name: Website
```
Comment (adamdmharvey) [inline packages/catalog-model/examples/acme/org.yaml:8]: Title, not name, ya?

File: packages/catalog-model/examples/acme/org.yaml [hunk 12]
```diff
@@ -3,6 +3,11 @@ kind: Group
...
+    - url: https://meta.wikimedia.org/wiki/
+      name: Intranet
```
Comment (adamdmharvey) [inline packages/catalog-model/examples/acme/org.yaml:10]: Title, not name, ya?

**Review by adamdmharvey — CHANGES_REQUESTED** [review]

File: packages/catalog-model/examples/acme/org.yaml (hunk 11 above)
Comment (andrewthauer) [reply packages/catalog-model/examples/acme/org.yaml:8]: 😢 Where is my head today ...

Comment (adamdmharvey) [reply packages/catalog-model/examples/acme/org.yaml:8]: Pixies! https://www.youtube.com/watch?v=49FB9hhoO6c

File: packages/catalog-model/src/validation/CommonValidatorFunctions.ts [hunk 13]
```diff
@@ -92,4 +92,35 @@ export class CommonValidatorFunctions {
...
+   * @param value The value to check
+   */
+  static isValidUrl(value: unknown): boolean {
+    if (typeof value !== 'string') {
+      return false;
+    }
+
+    try {
+      const url = new URL(value);
+      return !!url;
```
[the remark this answers is not in the document]
Comment (andrewthauer) [reply packages/catalog-model/src/validation/CommonValidatorFunctions.ts:108]: This was to work around linting errors. Eslint doesn't like `new URL(value)` (no-new violation). Assigning it to a variable and returning true means the variable isn't used 😞

File: packages/catalog-model/examples/apis/swapi-graphql.yaml (hunk 3 above)
Comment (andrewthauer) [reply packages/catalog-model/examples/apis/swapi-graphql.yaml]: Sorry already reverted. Happy to change it back in another PR for another example.

File: packages/catalog-model/src/validation/types.ts (hunk 9 above)
Comment (andrewthauer) [reply packages/catalog-model/src/validation/types.ts:29]: Yep, that seems to make sense to me. This is my first time making modifications to the catalog model so wasn't entirely sure what should and shouldn't be included.

**Review by adamdmharvey — APPROVED** [review]
Good re: my recommended changes, thanks @andrewthauer 👍

File: packages/catalog-model/src/validation/CommonValidatorFunctions.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (freben) [reply packages/catalog-model/src/validation/CommonValidatorFunctions.ts:108]: I see. Anyway I think using the URL constructor like this to check was the right decision, doing it with a regex or so is just bound to be less robust.

[the remark this answers is not in the document]
Comment (freben) [reply packages/catalog-model/src/validation/CommonValidatorFunctions.ts:108]: Oh and by the way - since this was a perfectly legitimate use of a bare `new`, it would be fine to just silence the lint error with a comment as well.

File: packages/catalog-model/src/validation/types.ts (hunk 9 above)
Comment (freben) [reply packages/catalog-model/src/validation/types.ts]: Nah we don't. Could be improved wholesale across the project some time. We're switching to json-schema anyway so it'll be easy to slap checking regexes in there later 

https://github.com/backstage/backstage/pull/4277

(don't worry BTW about coordination, I'll let you know if we will collide, but I suspect you'll go in first anyway...)

Comment (freben) [reply packages/catalog-model/src/validation/types.ts]: Or... Did I read that question right? :)

Comment (andrewthauer) [reply packages/catalog-model/src/validation/types.ts]: Either way, #4277 doesn't seem to have many conflicts at all from my skim of the PR.
